libc/string_cmp.c

```diff
@@ -0,0 +1,238 @@
+/*
+ * Phoenix-RTOS
+ *
+ * POSIX.1-2017 standard library functions tests
+ * HEADER:
+ *    - string.h
+ * TESTED:
+ *    - memcmp()
+ *    - strcmp()
+ *    - strncmp()
+ *    - strcoll()
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Damian Modzelewski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+#include <limits.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <limits.h>
+#include <unity_fixture.h>
+
+
+#define BUFF_SIZE 128
+
+static char asciiSet[BUFF_SIZE] = { 0 },
+			nonZeroAsciiSet[BUFF_SIZE] = { 0 },
+			empty[BUFF_SIZE] = { 0 },
+			separated[24] = { 0, 0, 0, 0, 0, 0, 'T', 'E', 'S', 'T', 0, 0, 0 };
+
+
+TEST_GROUP(string_cmp);
+
+
+TEST_SETUP(string_cmp)
+{
+}
+
+
+TEST_TEAR_DOWN(string_cmp)
+{
+}
+
+
+/*
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+*/
+
+
+TEST(string_cmp, memcmp_basic)
+{
+	TEST_ASSERT_EQUAL_INT(0, memcmp(asciiSet, asciiSet, BUFF_SIZE));
+	TEST_ASSERT_LESS_THAN_INT(0, memcmp(asciiSet, nonZeroAsciiSet, BUFF_SIZE));
+	TEST_ASSERT_GREATER_THAN_INT(0, memcmp(nonZeroAsciiSet, asciiSet, BUFF_SIZE));
+	TEST_ASSERT_EQUAL_INT(0, errno);
+}
+
+
+TEST(string_cmp, memcmp_empty)
+{
+	TEST_ASSERT_LESS_THAN_INT(0, memcmp(empty, asciiSet, BUFF_SIZE));
+	TEST_ASSERT_LESS_THAN_INT(0, memcmp(empty, nonZeroAsciiSet, BUFF_SIZE));
+	TEST_ASSERT_GREATER_THAN_INT(0, memcmp(asciiSet, empty, BUFF_SIZE));
+	TEST_ASSERT_EQUAL_INT(0, memcmp(empty, empty, BUFF_SIZE));
+
+	/* memory cmp is not sensitive for NUL characters */
+	TEST_ASSERT_NOT_EQUAL_INT(0, memcmp(empty, separated, sizeof(separated)));
+	TEST_ASSERT_EQUAL_INT(0, errno);
+}
+
+
+TEST(string_cmp, memcmp_sizes)
+{
+	int i;
+
+	TEST_ASSERT_LESS_THAN_INT(0, memcmp(&asciiSet[BUFF_SIZE / 2], &nonZeroAsciiSet[BUFF_SIZE / 2], BUFF_SIZE / 2));
+	TEST_ASSERT_GREATER_THAN_INT(0, memcmp(&asciiSet[sizeof(asciiSet) - 1], &nonZeroAsciiSet[sizeof(nonZeroAsciiSet) - 1], 1));
+	TEST_ASSERT_LESS_THAN_INT(0, memcmp(&nonZeroAsciiSet[BUFF_SIZE - 1], &asciiSet[BUFF_SIZE - 1], 1));
+	TEST_ASSERT_EQUAL_INT(0, memcmp(&nonZeroAsciiSet[sizeof(nonZeroAsciiSet) - 1], &nonZeroAsciiSet[sizeof(nonZeroAsciiSet) - 1], BUFF_SIZE - 1));
+	TEST_ASSERT_EQUAL_INT(0, memcmp(asciiSet, nonZeroAsciiSet, 0));
+
+	for (i = 1; i < sizeof(asciiSet); i++) {
+		TEST_ASSERT_LESS_THAN_INT(0, memcmp(asciiSet, nonZeroAsciiSet, i));
+	}
+
+	TEST_ASSERT_EQUAL_INT(0, errno);
+}
+
+
+/*
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+*/
+
+
+TEST(string_cmp, strncmp_basic)
+{
+	TEST_ASSERT_EQUAL_INT(0, strncmp(asciiSet, asciiSet, BUFF_SIZE));
```

Comment:
It was intentional as you can see in one line after this we used asciiSet and nonZeroAsciiSet.
This case supposes to force *cmp to compare elements that are the same and we try to catch if it will assign data correctly.